packages/@ember/-internals/glimmer/lib/components/input.ts

```diff
@@ -112,9 +269,510 @@ import InternalComponent from './internal';
+        'Did you mean `<Input @type="checkbox" @checked={{...}} />`?',
+      untrack(
+        () =>
+          this.args.checked !== undefined ||
+          this.args.value === undefined ||
+          typeof valueForRef(this.args.value) === 'string'
+      ),
+      { id: 'ember.built-in-components.input-checkbox-value' }
+    );
+
+    this._checked.set(checked);
+  }
+
+  _value = valueFrom(this.args.value);
+
+  get value(): unknown {
+    return this._value.get();
+  }
+
+  set value(value: unknown) {
+    this._value.set(value);
+  }
+
+  @action checkedDidChange(event: Event): void {
+    this.checked = this.elementFor(event).checked;
+  }
+
+  @action valueDidChange(event: Event): void {
+    this.value = this.valueFor(event);
+  }
+
+  @action change(event: Event): void {
+    if (this.isCheckbox) {
+      this.checkedDidChange(event);
+    } else {
+      this.valueDidChange(event);
+    }
+  }
+
+  @action input(event: Event): void {
+    if (!this.isCheckbox) {
+      this.valueDidChange(event);
+    }
+  }
+
+  @action keyUp(event: KeyboardEvent): void {
+    let value = this.valueFor(event);
+
+    switch (event.key) {
+      case 'Enter':
+        this.callbackFor('enter')(value, event);
+        this.callbackFor('insert-newline')(value, event);
+        break;
+
+      case 'Escape':
+        this.callbackFor('escape-press')(value, event);
+        break;
+    }
+  }
+
+  private elementFor(event: Event): HTMLInputElement {
+    assert(
+      '[BUG] Event target must be the <input> element',
+      event.target instanceof HTMLInputElement
+    );
+
+    return event.target;
+  }
+
+  private valueFor(event: Event): string {
+    return this.elementFor(event).value;
+  }
+
+  private callbackFor(type: string): (value: string, event: Event) => void {
+    let callback = this.arg(type);
+
+    if (callback) {
+      assert(
+        `The \`@${type}\` argument to the <Input> component must be a function`,
+        typeof callback === 'function'
+      );
+      return callback as (value: string, event: Event) => void;
+    } else {
+      return NOOP;
+    }
+  }
+}
+
+// Deprecated features
+if (EMBER_MODERNIZED_BUILT_IN_COMPONENTS) {
+  // Attribute bindings
+  {
+    let defineGetterForDeprecatedAttributeBinding = (
+      attribute: string,
+      argument = attribute
+    ): void => {
+      assert(
+        `[BUG] There is already a getter for _${argument} on Input`,
+        !Object.getOwnPropertyDescriptor(Input.prototype, `_${argument}`)
+      );
+
+      Object.defineProperty(Input.prototype, `_${argument}`, {
+        get(this: Input): unknown {
+          deprecate(
+            `Passing the \`@${argument}\` argument to <Input> is deprecated. ` +
+              `Instead, please pass the attribute directly, i.e. \`<Input ${attribute}={{...}} />\` ` +
+              `instead of \`<Input @${argument}={{...}} />\` or \`{{input ${argument}=...}}\`.`,
+            true /* TODO !(argument in this.args) */,
```

Comment:
Actually enabling all these deprecations require fixing _lots_ of tests, so leave them disabled for now.